plugins/scaleflux/sfx-nvme.c

```diff
@@ -0,0 +1,877 @@
+
+struct nvme_capacity_info {
+    __u64 lba_sec_sz;
+    __u64 phy_sec_sz;
+    __u64 used_space;
+    __u64 free_space;
+};
+#pragma pack(push,1)
+struct nvme_additional_smart_log_item {
+    uint8_t            key;
+    uint8_t            _kp[2];
+    uint8_t            norm;
+    uint8_t            _np;
+    union {
+        uint8_t        raw[6];
+        struct wear_level {
+            uint16_t    min;
+            uint16_t    max;
+            uint16_t    avg;
+        } wear_level ;
+        struct thermal_throttle {
+            uint8_t    pct;
+            uint32_t    count;
+        } thermal_throttle;
+    };
+    uint8_t            _rp;
+};
+#pragma pack(pop)
+
+struct nvme_additional_smart_log {
+    struct nvme_additional_smart_log_item    program_fail_cnt;
+    struct nvme_additional_smart_log_item    erase_fail_cnt;
+    struct nvme_additional_smart_log_item    wear_leveling_cnt;
+    struct nvme_additional_smart_log_item    e2e_err_cnt;
+    struct nvme_additional_smart_log_item    crc_err_cnt;
+    struct nvme_additional_smart_log_item    timed_workload_media_wear;
+    struct nvme_additional_smart_log_item    timed_workload_host_reads;
+    struct nvme_additional_smart_log_item    timed_workload_timer;
+    struct nvme_additional_smart_log_item    thermal_throttle_status;
+    struct nvme_additional_smart_log_item    retry_buffer_overflow_cnt;
+    struct nvme_additional_smart_log_item    pll_lock_loss_cnt;
+    struct nvme_additional_smart_log_item    nand_bytes_written;
+    struct nvme_additional_smart_log_item    host_bytes_written;
+    struct nvme_additional_smart_log_item    raid_recover_cnt; // errors which can be recovered by RAID
+    struct nvme_additional_smart_log_item    prog_timeout_cnt;
+    struct nvme_additional_smart_log_item    erase_timeout_cnt;
+    struct nvme_additional_smart_log_item    read_timeout_cnt;
+    struct nvme_additional_smart_log_item    read_ecc_cnt;//retry cnt
+};
+
+int nvme_change_cap(int fd, __u32 nsid, __u64 capacity)
+{
+    struct nvme_admin_cmd cmd = {
+    .opcode      = nvme_admin_change_cap,
+    .nsid	     = nsid,
+    };
+
+    cmd.cdw10 = (capacity & 0xffffffff);
+    cmd.cdw11 = (capacity >> 32);
+
+    return nvme_submit_passthru(fd, NVME_IOCTL_ADMIN_CMD,&cmd);
+}
+
+int nvme_sfx_set_features(int fd, __u32 nsid, __u32 fid, __u32 value)
+{
+    struct nvme_admin_cmd cmd = {
+    .opcode      = nvme_admin_sfx_set_features,
+    .nsid        = nsid,
+    .cdw10       = fid,
+    .cdw11       = value,
+    };
+
+    return nvme_submit_passthru(fd, NVME_IOCTL_ADMIN_CMD,&cmd);
+}
+
+int nvme_sfx_get_features(int fd, __u32 nsid, __u32 fid, __u32 *result)
+{
+    int err = 0;
+    struct nvme_admin_cmd cmd = {
+    .opcode      = nvme_admin_sfx_get_features,
+    .nsid        = nsid,
+    .cdw10       = fid,
+    };
+
+    err = nvme_submit_passthru(fd, NVME_IOCTL_ADMIN_CMD,&cmd);
+    if (!err && result) {
+        *result = cmd.result;
+    }
+    printf("%d %d %p\n", *result,
```

Comment:
Since 'result' is assigned only on success, you are potentially printing an uninitialized value. But the only caller prints it on success anyway, so I'm not sure why you're printing it here.

Reply:
will remove.